### ISSUE TITLE
SSL Fatal Exception Insufficient Security

### ISSUE THREAD

**Issue by mikenemat** [mikenemat/org.openhab.binding.plum#7]
Some people are seeing errors like this. Please post output of "java -version" to assist troubleshooting

11:00:51.201 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP
Request to IP 10.0.1.36 returned ERROR javax.net.ssl.SSLException: Received
fatal alert: insufficient_security
11:00:51.229 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP
Request to IP 10.0.1.27 returned ERROR javax.net.ssl.SSLException: Received
fatal alert: insufficient_security
11:00:51.307 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP
Request to IP 10.0.1.26 returned ERROR javax.net.ssl.SSLException: Received
fatal alert: insufficient_security
11:00:51.334 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP
Request to IP 10.0.1.25 returned ERROR javax.net.ssl.SSLException: Received
fatal alert: insufficient_security
11:00:51.409 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP
Request to IP 10.0.1.32 returned ERROR javax.net.ssl.SSLException: Received
fatal alert: insufficient_security
11:00:51.436 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP
Request to IP 10.0.1.31 returned ERROR javax.net.ssl.SSLException: Received
fatal alert: insufficient_security
11:00:51.466 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP
Request to IP 10.0.1.28 returned ERROR javax.net.ssl.SSLException: Received
fatal alert: insufficient_security
…

**Comment by billmans** [comment]
Hey Mike,

Here is the java version I am using.  I have not updated this in a while.

openHABMac:~ openhab$ java -version
openjdk version "1.8.0_162"
OpenJDK Runtime Environment (Zulu 8.27.0.7-macosx) (build 1.8.0_162-b01)
OpenJDK 64-Bit Server VM (Zulu 8.27.0.7-macosx) (build 25.162-b01, mixed mode)

**Comment by mikenemat** [comment]
Another request, can you send me the output of the following:

openssl s_client -host 10.0.1.28 -port 8443

**Comment by mikenemat** [comment]
Also:

jrunscript -e 'exit (javax.crypto.Cipher.getMaxAllowedKeyLength("RC5") >= 256 ? 0 : 1);'; echo $?

**Comment by mikenemat** [comment]
I am starting to suspect this is a variant of the problem described in https://community.openhab.org/t/loxone-binding-internal-error-enable-java-cryptography-unlimited-strength-see-binding-doc/49266/10 but I need more info since I am unable to replicate it locally :)

**Comment by mikenemat** [comment]
And one more:

wget https://confluence.atlassian.com/bitbucketserverkb/files/779171661/779171662/1/1414093373406/Ciphers.java
javac Ciphers.java
java Ciphers

**Comment by billmans** [comment]
openHABMac:~ openhab$ openssl s_client -host 10.0.1.28 -port 8443
CONNECTED(00000003)
depth=1 C = CH, O = Plum Inc., CN = plum.equipment
verify error:num=19:self signed certificate in certificate chain
verify return:0
---
Certificate chain
 0 s:/C=US/ST=Texas/L=Austin/O=Plum
   i:/C=CH/O=Plum Inc./CN=plum.equipment
 1 s:/C=CH/O=Plum Inc./CN=plum.equipment
   i:/C=CH/O=Plum Inc./CN=plum.equipment
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIC8zCCAdsCCQDpm4Vt6mKtYTANBgkqhkiG9w0BAQsFADA6MQswCQYDVQQGEwJD
SDESMBAGA1UEChMJUGx1bSBJbmMuMRcwFQYDVQQDEw5wbHVtLmVxdWlwbWVudDAe
Fw0xNzA5MTgxODQxMzlaFw0yMzA0MjgxODQxMzlaMD0xCzAJBgNVBAYTAlVTMQ4w
DAYDVQQIEwVUZXhhczEPMA0GA1UEBxMGQXVzdGluMQ0wCwYDVQQKEwRQbHVtMIIB
IjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA5KuIdjpSWCf/z8rulcqucAxo
0ggNhNaARpHvp7dwy9/8eFxDmAHzxFqbQReDIHHKRLcM/hkwDUFHZg7hp4ww35Hk
ctBuikeNrJTmYW8kwfelbjjAFf+RN1L0mD7oliG48zDJxps1PWy30gqSBEZYa4ld
vX71i+m6wkZE/ZSnYOD9q3u/hhXou8eWN7BYIFYEEOb3bircODsuSRZAeFPQbz+o
l2ZAVRfrNVdF1i8ZoLjUJJRc47bfTiglHfmreuM6LvMpoUP4Rjt3Xm+KtPel7Mx6
F1Bw7bJM7lRf8Kt1XWgQapqjlJ7Ey0ctdx5/Ci7yRBfPA11vAzjUQIaUcAjLgQID
AQABMA0GCSqGSIb3DQEBCwUAA4IBAQAiO+YVD8Afxqkgn3uq9PUcgqdo9xbYcKKn
bx3Nd86IjDYWFBfdZmY2k+BFZMgtB0DlT3fwyHk5nbQx0ZpfnIRBKuT9Huxhv102
ljtZCmJnhdSEHtKCuw/6d1I+HUBDHiVSVoZCToz/Aq3Z1hQ3dvl4Sbj52asWI9Px
cjfvqipM8kVDzvoYQ1/OABKHDAvfSquzf9am/4GaRfJ0BtS1k3aoexOAWMweeA5D
flBtyYSr+3Ncb5Jqd/Q7Ht8/0FETQdK7T0r74tvITBz2SMulnmKa8Vu4I+zSyl5g
PkyuJyc41IQ7pLLIswLS4rAVjdXq+ByxtNuBlrbZEAwWVZ4pRty6
-----END CERTIFICATE-----
subject=/C=US/ST=Texas/L=Austin/O=Plum
issuer=/C=CH/O=Plum Inc./CN=plum.equipment
---
No client certificate CA names sent
---
SSL handshake has read 1825 bytes and written 676 bytes
---
New, TLSv1/SSLv3, Cipher is AES256-SHA256
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : AES256-SHA256
    Session-ID:
B5A62DB94E33FDD11237ED8016E89A5DAF916D351CB8C83B74D6E9E47E52B553
    Session-ID-ctx:
    Master-Key:
94D10D2146212B6D1A8090003ACBB6114E278885FF9EE18C8B88C7352289BA4598C8ACC248DB851E59BDC2EABA83B7A3
    Start Time: 1557973622
    Timeout   : 300 (sec)
    Verify return code: 19 (self signed certificate in certificate chain)
---
closed

**Comment by billmans** [comment]
openHABMac:~ openhab$ jrunscript -e 'exit
(javax.crypto.Cipher.getMaxAllowedKeyLength("RC5") >= 256 ? 0 : 1);'; echo
$?
0

**Comment by billmans** [comment]
openHABMac:~ openhab$ curl -O
https://confluence.atlassian.com/bitbucketserverkb/files/779171661/779171662/1/1414093373406/Ciphers.java
  % Total    % Received % Xferd  Average Speed   Time    Time     Time
Current
                                 Dload  Upload   Total   Spent    Left
Speed
100  1143  100  1143    0     0   2695      0 --:--:-- --:--:-- --:--:--
2695
openHABMac:~ openhab$ javac Ciphers.java
Note: Ciphers.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
openHABMac:~ openhab$ java Ciphers
Default Cipher
* SSL_DHE_DSS_WITH_3DES_EDE_CBC_SHA
  SSL_DHE_DSS_WITH_DES_CBC_SHA
* SSL_DHE_RSA_WITH_3DES_EDE_CBC_SHA
  SSL_DHE_RSA_WITH_DES_CBC_SHA
  SSL_DH_anon_WITH_3DES_EDE_CBC_SHA
  SSL_DH_anon_WITH_DES_CBC_SHA
* SSL_RSA_WITH_3DES_EDE_CBC_SHA
  SSL_RSA_WITH_DES_CBC_SHA
  SSL_RSA_WITH_NULL_MD5
  SSL_RSA_WITH_NULL_SHA
* TLS_DHE_DSS_WITH_AES_128_CBC_SHA
* TLS_DHE_DSS_WITH_AES_128_CBC_SHA256
* TLS_DHE_DSS_WITH_AES_128_GCM_SHA256
* TLS_DHE_DSS_WITH_AES_256_CBC_SHA
* TLS_DHE_DSS_WITH_AES_256_CBC_SHA256
* TLS_DHE_DSS_WITH_AES_256_GCM_SHA384
* TLS_DHE_RSA_WITH_AES_128_CBC_SHA
* TLS_DHE_RSA_WITH_AES_128_CBC_SHA256
* TLS_DHE_RSA_WITH_AES_128_GCM_SHA256
* TLS_DHE_RSA_WITH_AES_256_CBC_SHA
* TLS_DHE_RSA_WITH_AES_256_CBC_SHA256
* TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
  TLS_DH_anon_WITH_AES_128_CBC_SHA
  TLS_DH_anon_WITH_AES_128_CBC_SHA256
  TLS_DH_anon_WITH_AES_128_GCM_SHA256
  TLS_DH_anon_WITH_AES_256_CBC_SHA
  TLS_DH_anon_WITH_AES_256_CBC_SHA256
  TLS_DH_anon_WITH_AES_256_GCM_SHA384
* TLS_ECDHE_ECDSA_WITH_3DES_EDE_CBC_SHA
* TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
* TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
* TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
* TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
* TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
* TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
  TLS_ECDHE_ECDSA_WITH_NULL_SHA
* TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA
* TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
* TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
* TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
* TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384
* TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
  TLS_ECDHE_RSA_WITH_NULL_SHA
* TLS_ECDH_ECDSA_WITH_3DES_EDE_CBC_SHA
* TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA
* TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256
* TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256
* TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA
* TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA384
* TLS_ECDH_ECDSA_WITH_AES_256_GCM_SHA384
  TLS_ECDH_ECDSA_WITH_NULL_SHA
* TLS_ECDH_RSA_WITH_3DES_EDE_CBC_SHA
* TLS_ECDH_RSA_WITH_AES_128_CBC_SHA
* TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256
* TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256
* TLS_ECDH_RSA_WITH_AES_256_CBC_SHA
* TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384
* TLS_ECDH_RSA_WITH_AES_256_GCM_SHA384
  TLS_ECDH_RSA_WITH_NULL_SHA
  TLS_ECDH_anon_WITH_3DES_EDE_CBC_SHA
  TLS_ECDH_anon_WITH_AES_128_CBC_SHA
  TLS_ECDH_anon_WITH_AES_256_CBC_SHA
  TLS_ECDH_anon_WITH_NULL_SHA
* TLS_EMPTY_RENEGOTIATION_INFO_SCSV
  TLS_KRB5_WITH_3DES_EDE_CBC_MD5
  TLS_KRB5_WITH_3DES_EDE_CBC_SHA
  TLS_KRB5_WITH_DES_CBC_MD5
  TLS_KRB5_WITH_DES_CBC_SHA
* TLS_RSA_WITH_AES_128_CBC_SHA
* TLS_RSA_WITH_AES_128_CBC_SHA256
* TLS_RSA_WITH_AES_128_GCM_SHA256
* TLS_RSA_WITH_AES_256_CBC_SHA
* TLS_RSA_WITH_AES_256_CBC_SHA256
* TLS_RSA_WITH_AES_256_GCM_SHA384
  TLS_RSA_WITH_NULL_SHA256
openHABMac:~ openhab$

**Comment by billmans** [comment]
Hi Mike,

I followed the instructions on the following thread and added the newer Jar files.  Rebooted and does not seem to make any difference.

https://community.openhab.org/t/loxone-binding-internal-error-enable-java-cryptography-unlimited-strength-see-binding-doc/49266/7


Still seeing 

17:04:34.857 [ERROR] [g.plum.internal.PlumTCPStreamListener] - Plum TCP Stream error for IP 10.0.1.30 Message: Operation timed out (Connection timed out)
17:04:41.820 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP Request to IP 10.0.1.34 returned ERROR javax.net.ssl.SSLException: Received fatal alert: insufficient_security
17:04:46.824 [WARN ] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP Request to IP 10.0.1.30 timed out. Check IP or reboot Lightpad
17:04:46.855 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP Request to IP 10.0.1.29 returned ERROR javax.net.ssl.SSLException: Received fatal alert: insufficient_security
17:04:46.884 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP Request to IP 10.0.1.36 returned ERROR javax.net.ssl.SSLException: Received fatal alert: insufficient_security
17:04:46.915 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP Request to IP 10.0.1.27 returned ERROR javax.net.ssl.SSLException: Received fatal alert: insufficient_security
17:04:46.951 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP Request to IP 10.0.1.26 returned ERROR javax.net.ssl.SSLException: Received fatal alert: insufficient_security
17:04:46.981 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP Request to IP 10.0.1.25 returned ERROR javax.net.ssl.SSLException: Received fatal alert: insufficient_security
17:04:47.011 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP Request to IP 10.0.1.32 returned ERROR javax.net.ssl.SSLException: Received fatal alert: insufficient_security
17:04:47.039 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP Request to IP 10.0.1.31 returned ERROR javax.net.ssl.SSLException: Received fatal alert: insufficient_security
17:04:47.069 [ERROR] [penhab.binding.plum.PlumActiveBinding] - Plum HTTP Request to IP 10.0.1.28 returned ERROR javax.net.ssl.SSLException: Received fatal alert: insufficient_security

**Comment by mikenemat** [comment]
Alright. I've made some progress.  Here's a new JAR to try. I'm not even going to begin to try and understand the issue. Long story short, it appears that there is some preferred SSL/TLS cipher that some java versions use by default (maybe due to an OS update?) which the Plum lightpad claims to support but actually doesn't, so the SSL auth fails. 

https://drive.google.com/open?id=135A5aBDuazKrvWixcilzUuFfzaP3nU6z

Let me know if this works. Remember to uninstall the old plum JAR by doing
bundle:list (find the ID)
bundle:uninstall ID
rm the old JAR In the addons folder
restart openhab
copy the new JAR to the addons folder.

**Comment by dome2048** [comment]
This new jar is fixing my issue (as reported on the [openHAB forums](https://community.openhab.org/t/trouble-with-plum-java-permissions/71567) linked below).

Thanks a ton @mikenemat !

**Comment by mikenemat** [comment]
Once I get a bit more feedback and have tested it in my own house for a week or two, I will check the code changes into the git repo. For now, everyone use the google drive link plz :)

**Comment by billmans** [comment]
Hi Mike,

Finally had a chance to test this morning, looks good I can control my plum
dimmers again and the errors seem to be gone.

Thanks again so much.  Will keep an eye and let you know if I see any
issues.

I wonder how long these will continue to work?  I believe your plugin
communicates locally, but do you think they will stop working if they shut
down the server?  What about certificates etc on the lightpads, when they
expire?

I still have 3 or 4 I have not even installed, I guess without the server I
would not be able to provision them and use them at all. (guess I should at
least see if I can get that done!)

Thanks
Bill

**Comment by mikenemat** [comment]
Hi Bill, I know it's been a while, just saw your message now. My plugin ignores certs on the lightpads and communicates locally. As long as you don't reset them and keep the LLID values the same, it should work forever.

The retrieval of LLID values using the plum-probe.py tool in my other repo does require online services to function, though I suspect it could be made to work entirely locally with some effort. It wouldn't be able to correlate rooms and names, but it would be able to enumerate the lightpads. 

I've heard some talk around the internet of provisioning not working anymore, can't say that would be surprising given it was dependent on their cloud services. I haven't tried. 

Last week I removed all of my lightpads as well as the 3 spare units I had and discarded them. I was having hardware issues with 4 of my 18 units. Good riddance and lesson learned. Beautiful hardware and still nothing quite like it, but the software and reliability just wasn't there. 

I replaced all my Plum hardware with Lutron Caseta, and I have a lot of Insteon hardware too (but not dimmers...I find their dimmers flicker badly with LED lighting).  Not nearly as flashy, but utterly reliable. I have tried over a dozen different smart dimmers before settling on Caseta. I just wish it was cheaper. Their fully functional universal dimmer is $99 USD. The cheaper one without a neutral wire that's $40 or $50 causes a lot of my LED pot lights to buzz.
